### PR TITLE
Internal Java Solution for Bovine Genomics (Silver)

### PR DESCRIPTION
**Pull request by brad182**

### CODE REVIEW

File: solutions/usaco-739.mdx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  
+```java
+import java.io.*;
+import java.util.StringTokenizer;
+
+public class BovineGenomics {
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio();
+		//BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		//PrintWriter pw = new PrintWriter(System.out);
```
Comment (maggie-j-liu) [inline solutions/usaco-739.mdx:30]: maybe just delete these

File: solutions/usaco-739.mdx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+This implementation uses sets, which has not been covered yet.
+</Warning>
+  
+```java
+import java.io.*;
+import java.util.StringTokenizer;
+
+public class BovineGenomics {
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio();
```
Comment (maggie-j-liu) [inline solutions/usaco-739.mdx:28]: forgot filename?

File: solutions/usaco-739.mdx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+## Solution
+
+**Time Complexity** $\mathcal{O}(N^3)$
+
+<LanguageSection>
+  
+<JavaSection>
+  
+<Warning>
+This implementation uses sets, which has not been covered yet.
```
Comment (maggie-j-liu) [inline solutions/usaco-739.mdx:19]:
```suggestion
This implementation uses sets, which have not been covered yet.
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-739.mdx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+						// get the base 4 number
+						int total = spotted[x][a] * 16 + spotted[x][b] * 4 + spotted[x][c] * 1;
+						visited[total] = true;
+					}
+
+					boolean add = true;
+					for (int x = 0; x < n; x++) {
+						// get the base 4 number
+						int total = plain[x][a] * 16 + plain[x][b] * 4 + plain[x][c] * 1;
+						if (visited[total]) {  // can't explain, don't add
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-739.mdx:97]:
```suggestion
						if (visited[total]) {  // FJ can't explain with this combination of 3, so don't add
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/usaco-739.mdx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+		//BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		//PrintWriter pw = new PrintWriter(System.out);
+
+		int n = io.nextInt();
+		int m = io.nextInt();
+
+		int[][] spotted = new int[n][m];
+		for (int x = 0; x < n; x++) {  // read spotted
+			StringBuilder line = new StringBuilder();
+			line.append(io.next());
```
[the remark this answers is not in the document]
Comment (maggie-j-liu) [reply solutions/usaco-739.mdx:38]: kattio does have a next function (or at least the modified kattio that we use)
image

[the remark this answers is not in the document]
Comment (Aadit-Ambadkar) [reply solutions/usaco-739.mdx:38]: Hmm ok. The Kattio I have doesn't have it.

File: solutions/usaco-739.mdx [hunk 6]
```diff
@@ -0,0 +1,142 @@
+---
+id: usaco-739
+source: USACO Silver 2017 US Open 
+title: Bovine Genomics
+author: Brad Ma
+---
+
+[Official Editorial (C++)](http://www.usaco.org/current/data/sol_cownomics_silver_open17.html)
+
+## Solution
```
Comment (nathangong) [inline solutions/usaco-739.mdx:10]:
```suggestion
## Implementation
```

File: solutions/usaco-739.mdx [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+		Kattio io = new Kattio();
+		//BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		//PrintWriter pw = new PrintWriter(System.out);
+
+		int n = io.nextInt();
+		int m = io.nextInt();
+
+		int[][] spotted = new int[n][m];
+		for (int x = 0; x < n; x++) {  // read spotted
+			StringBuilder line = new StringBuilder();
```
Comment (nathangong) [inline solutions/usaco-739.mdx:37]: Why are you using a stringbuilder if you're only appending to it once?

File: solutions/usaco-739.mdx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+				}
+				else {
+					spotted[x][y] = 3;
+				}
+			}
+		}
+
+		int[][] plain = new int[n][m];
+		for (int x = 0; x < n; x++) {  // read plain
+			StringBuilder line = new StringBuilder();
```
Comment (nathangong) [inline solutions/usaco-739.mdx:61]: Same here, no need for a stringbuilder

File: solutions/usaco-739.mdx [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+		int count = 0;
+
+		for (int a = 0; a < m; a++) {  // loop through every possible group of 3
+			for (int b = a + 1; b < m; b++) {
+				for (int c = b + 1; c < m; c++) {
+					boolean[] visited = new boolean[193];  // 64 * 3 + 1
+					for (int x = 0; x < n; x++) {
+						// get the base 4 number
+						int total = spotted[x][a] * 16 + spotted[x][b] * 4 + spotted[x][c] * 1;
```
Comment (nathangong) [inline solutions/usaco-739.mdx:89]: I think this line is over 80 chars

File: solutions/usaco-739.mdx (hunk 3 above)
Comment (nathangong) [reply solutions/usaco-739.mdx:19]: Wait I dont even think this implementation uses sets

File: solutions/usaco-739.mdx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+				else {
+					spotted[x][y] = 3;
+				}
+			}
+		}
+
+		int[][] plain = new int[n][m];
+		for (int x = 0; x < n; x++) {  // read plain
+			StringBuilder line = new StringBuilder();
+			line.append(io.next());
```
[the remark this answers is not in the document]
Comment (nathangong) [reply solutions/usaco-739.mdx:62]: wait i'm pretty sure the kattio from usaco.guide uses `.next()` instead of `.getWord()`

**Comment by SansPapyrus683** [comment]
i'll review when everyone else's stuff has been resolved

File: solutions/usaco-739.mdx (hunk 9 above)
Comment (brad182) [reply solutions/usaco-739.mdx:89]: you are allowed to barely go over on a few lines

File: solutions/usaco-739.mdx (hunk 7 above)
Comment (brad182) [reply solutions/usaco-739.mdx:37]: good practice to use string builders on string problems, .charAt can be expensive

Comment (SansPapyrus683) [reply solutions/usaco-739.mdx:37]: .charAt has never failed me?

File: solutions/usaco-739.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply solutions/usaco-739.mdx:87]: image
"barely"

File: solutions/usaco-739.mdx [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+			for (int y = 0; y < m; y++) {
+				// A -> 1, C -> 4, T -> 16, G -> 64
```
Comment (SansPapyrus683) [inline solutions/usaco-739.mdx:63]: maybe explain what these numbers mean

File: solutions/usaco-739.mdx (hunk 7 above)
Comment (nathangong) [reply solutions/usaco-739.mdx:35]: yeah I don't think the charAt method in strings causes Java to TLE, normally it's i/o

File: solutions/usaco-739.mdx (hunk 11 above)
Comment (nathangong) [reply solutions/usaco-739.mdx:63]: also, it doesn't seem like you're using the numbers here (if you are, it's not obvious to the reader)

File: solutions/usaco-739.mdx [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+						// get the base 4 number
+						int total = spotted[x][a] * 16 + spotted[x][b] * 4 + spotted[x][c] * 1;
+						visited[total] = true;
+					}
+
+					boolean add = true;
+					for (int x = 0; x < n; x++) {
+						// get the base 4 number
+						int total = plain[x][a] * 16 + plain[x][b] * 4 + plain[x][c] * 1;
+						if (visited[total]) {  // FJ can't explain with this combination of 3, so don't add
```
Comment (maggie-j-liu) [inline solutions/usaco-739.mdx:95]: these two lines are too long (as well as the one nathan mentioned above)

**Review by ryanchou-dev — COMMENTED** [review]
(optional)

File: solutions/usaco-739.mdx [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+				}
+				else if (line.charAt(y) == 'C') {
+					spotted[x][y] = 1;
+				}
+				else if (line.charAt(y) == 'T') {
+					spotted[x][y] = 2;
+				}
+				else {
+					spotted[x][y] = 3;
+				}
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-739.mdx:53]:
```suggestion
				if (line.charAt(y) == 'A') {
					spotted[x][y] = 0;
				} else if (line.charAt(y) == 'C') {
					spotted[x][y] = 1;
				} else if (line.charAt(y) == 'T') {
					spotted[x][y] = 2;
				} else {
					spotted[x][y] = 3;
				}
```

File: solutions/usaco-739.mdx [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+public class BovineGenomics {
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio("cownomics");
+	
+		int n = io.nextInt();
+		int m = io.nextInt();
+
+		int[][] spotted = new int[n][m];
+		for (int x = 0; x < n; x++) {  // read spotted
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-739.mdx:34]:
```suggestion
		for (int x = 0; x < n; x++) {  // read spotted array
```

File: solutions/usaco-739.mdx [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+					spotted[x][y] = 2;
+				}
+				else {
+					spotted[x][y] = 3;
+				}
+			}
+		}
+
+		int[][] plain = new int[n][m];
+		for (int x = 0; x < n; x++) {  // read plain
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-739.mdx:58]:
```suggestion
		for (int x = 0; x < n; x++) {  // read plain array
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/usaco-739.mdx (hunk 9 above)
Comment (brad182) [reply solutions/usaco-739.mdx:87]: less than +20% is barely

File: solutions/usaco-739.mdx (hunk 13 above)
Comment (brad182) [reply solutions/usaco-739.mdx:53]: that is one of those things that everybody does differently and qualifies as an optional change

File: solutions/usaco-739.mdx (hunk 11 above)
Comment (brad182) [reply solutions/usaco-739.mdx:63]: will multiply later on, added this as a comment

File: solutions/usaco-739.mdx (hunk 7 above)
Comment (Aadit-Ambadkar) [reply solutions/usaco-739.mdx:35]: .charAt is O(1) for Strings as well. [See this](https://stackoverflow.com/a/6461439)

File: solutions/usaco-739.mdx (hunk 13 above)
Comment (Aadit-Ambadkar) [reply solutions/usaco-739.mdx:59]: IDK if USACO Guide uses this, but according to Oracle, [this ](https://www.oracle.com/java/technologies/javase/codeconventions-statements.html#449)is the proper formatting:

File: solutions/usaco-739.mdx [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+				}
+				else if (line.charAt(y) == 'C') {
+					plain[x][y] = 1;
+				}
+				else if (line.charAt(y) == 'T') {
+					plain[x][y] = 2;
+				}
+				else {
+					plain[x][y] = 3;
+				}
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-739.mdx:80]:
```suggestion
				if (line.charAt(y) == 'A') {
					plain[x][y] = 0;
				} else if (line.charAt(y) == 'C') {
					plain[x][y] = 1;
				} else if (line.charAt(y) == 'T') {
					plain[x][y] = 2;
				} else {
					plain[x][y] = 3;
				}
```
Same as above

File: solutions/usaco-739.mdx (hunk 12 above)
Comment (Aadit-Ambadkar) [reply solutions/usaco-739.mdx:101]:
```suggestion
						int total = plain[x][a] * 16 + plain[x][b] * 4 + plain[x][c] * 1;
						// FJ can't explain with this combination of 3, so don't add
						if (visited[total]) {
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/usaco-739.mdx (hunk 16 above)
Comment (brad182) [reply solutions/usaco-739.mdx:80]: this is optional for code on the guide

File: solutions/usaco-739.mdx [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+## Implementation
+
+**Time Complexity** $\mathcal{O}(N^3)$
+
+<LanguageSection>
+  
+<JavaSection>
+  
+<Warning>
+This implementation uses the set data structure, which has not been covered yet. See [Introduction to Sets/Maps](https://usaco.guide/bronze/intro-sets) for more details about sets.
```
[the remark this answers is not in the document]
Comment (brad182) [reply solutions/usaco-739.mdx:19]: no

File: solutions/usaco-739.mdx (hunk 8 above)
Comment (brad182) [reply solutions/usaco-739.mdx:65]: you don't know about constants and always good to use string builders for string problems

File: solutions/usaco-739.mdx (hunk 7 above)
Comment (brad182) [reply solutions/usaco-739.mdx:35]: good practice to use string builders and there are constants

File: solutions/usaco-739.mdx (hunk 9 above)
Comment (brad182) [reply solutions/usaco-739.mdx]: changed

**Review by nwatx — APPROVED** [review]

**Review by Aadit-Ambadkar — APPROVED** [review]
I'll approve, but I believe that you shouldn't use an SB, and instead use a String. It is mainly stylistic though.

**Review by ar492 — APPROVED** [review]

File: solutions/usaco-739.mdx [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (ryanchou-dev) [inline solutions/usaco-739.mdx:149]: why are there so many blank lines?

**Review by ryanchou-dev — APPROVED** [review]
:D

File: solutions/usaco-739.mdx (hunk 18 above)
Comment (brad182) [reply solutions/usaco-739.mdx:149]: lol werid i'll change

**Comment by brad182** [comment]
merge?

**Review by nathangong — APPROVED** [review]

File: solutions/usaco-739.mdx (hunk 17 above)
[the remark this answers is not in the document]
Comment (dongliu0426) [reply solutions/usaco-739.mdx:19]: remove this? idt your solution uses the set data-structure.

**Review by dongliu0426 — CHANGES_REQUESTED** [review]
https://github.com/cpinitiative/usaco-guide/pull/2088#discussion_r711669450

File: solutions/usaco-739.mdx (hunk 17 above)
[the remark this answers is not in the document]
Comment (brad182) [reply solutions/usaco-739.mdx]: fixed, another implementation i did without base 4 used sets

**Review by dongliu0426 — APPROVED** [review]

**Pull request merged**